src/tools/pkcs15-tool.c

```diff
@@ -1515,11 +1515,47 @@ static int read_and_cache_file(const sc_path_t *path)
 		printf("out of memory!");
 		return -1;
 	}
-	r = sc_read_binary(card, 0, buf, size, 0);
-	if (r < 0) {
-		fprintf(stderr, "sc_read_binary() failed: %s\n", sc_strerror(r));
-		free(buf);
-		return -1;
+	if (tfile->ef_structure == SC_FILE_EF_LINEAR_VARIABLE_TLV) {
```

Comment:
On Thu, 05 Feb 2015 18:10:16 -0800
Frank Morgner <notifications@github.com> wrote:

> > @@ -1515,11 +1515,47 @@ static int read_and_cache_file(const sc_path_t *path)
> >  		printf("out of memory!");
> >  		return -1;
> >  	}
> > -	r = sc_read_binary(card, 0, buf, size, 0);
> > -	if (r < 0) {
> > -		fprintf(stderr, "sc_read_binary() failed: %s\n", sc_strerror(r));
> > -		free(buf);
> > -		return -1;
> > +	if (tfile->ef_structure == SC_FILE_EF_LINEAR_VARIABLE_TLV) {
> 
> some cards don't return metadata about the files so there is no way of knowing whether its a `SC_FILE_EF_LINEAR_VARIABLE_TLV`. This change would render pkcs15-tool unusable for many cards.
> 

I do not understand that argument. The code I was changing is only hit during
pkcs15-tool -L  (hence the commit message)
What does it make worse?
We have three cases here:
1. we know we have a binary file (which can be read using sc_read_binary())
2. we know we have a record based file
3. we do not know what kind of file we have.

without the patch we handle 1. correctly and assume we have a binary file in 3.
With the patch we handle 1. and 2. correctly and assume we have a binary file in 3.